### PR TITLE
Landlord submission form

### LINKED ISSUE #17: Simple Landlord Form: Form Page
# Overview

### Landlord Form

Implement just the landlord form page ([Figma](https://www.figma.com/file/06vEDN5gtlHlC5hxNXMb4g/USHS-Designs?node-id=339%3A8096&mode=dev)). It will be helpful to first create components (e.g. inputs, dropdowns, checkboxes, radio buttons, etc) for all the form elements. You can route it to the `/list` path in the app.

Because this is a big form, I recommend implementing this using [React Hook Form](https://github.com/react-hook-form/react-hook-form). This library makes developing forms with a lot of state super easy. It will also make things much easier for us in the future when we need to prefill this form with data (e.g. for editing a unit). An extensive example from a past TSE project of a large form developed using React Hook Form (and styled-components!) can be found [here](https://github.com/TritonSE/TAP-Foster-Dog-Matching/blob/main/client/src/components/FosterApplication.js). I'd recommend checking this out and familiarizing yourself with how it works. Don't be afraid to steal as much code (including any form element components!) as relevant for this task. In the example, components that are designed to work with React Hook Form start with `Controlled` (e.g. `ControlledInput`).

### Unit API

Create `frontend/src/api/units.ts` and create a function `createUnit(...)` that sends a request to `POST /api/units` sending all the relevant fields to create a new unit. You can follow `frontend/src/api/users.ts` as an example. It will also be helpful to create a `UnitCreateRequest` data type representing all the fields needed for the request. 

### Functionality

Integrate the frontend with the backend. When the "Submit" button is pressed, call the `createUnit` function from above, passing in the relevant form fields using `handleSubmit` from React Hook Form. Then, show the confirmation screen.

## Backend

N/A

## Acceptance Criteria
- [x] Landlord Page (`/list`) implemented
- [x] `frontend/src/api/unit.ts` and `createUnit` function implemented
- [x] End-to-end functionality works (filling out the form and clicking Submit will create a Unit object in the database)

### PR DESCRIPTION
**Pull request by brentonmdunn**
## Tracking Info

Resolves #17

## Changes

- Add all components for form
- Performed validation on input

## Testing

- Tested all of the inputs to validate that the radio and checkboxes worked.

## Confirmation of Change

image

image

image

image

image

### CODE REVIEW

**Review by petabite — CHANGES_REQUESTED** [review]
Nice work! 💯 This is looking really solid so far!

I demoed this during the team meeting and we noticed a few things:

- [x] Can we make the "Apartment/Suite" field not mandatory?
- [x] Increase spacing between the text box and n/a for the "Application Fee Cost" field
- [x] The "Additional fields to be entered by housing locator" section heading and label text should be a different color to highlight to HLs that they need to fill it out (see designs)
- [x] The "Additional Rules/Regulations" section has a bug where selecting some options and then selecting N/A does nothing but if you click it again, it fills in the radio button.
- [x] The Additional comments text box for the landlord is missing
- [x] Was not in the designs before, but can we turn the two comments text boxes (for landlord and HL) into textareas since it is possible they could could be typing in a lot of text?
- [x] The various number inputs (e.g. application fee cost, phone number, rent per month, security deposit, etc) should only allow numbers to be inputed. Currently, typing in letters is allowed
- [x] Was not in the designs before: For the "Select All Appliances that Apply or N/A" heading, can we delete the "or N/A" part?
- [x] Was not in the designs before: For "Payment/Renting Criteria" can we remove the “(select ALL that apply)” text
- [x] Is it possible to max the year part of the Date Available input to 4 digits (i.e. set the max year to 9999 or something)?
- [x] It looks like there are a few type errors in the `src/components/ListingFormComponents.tsx` file

I know this is a lot 😓  Please let me know if you have any questions about anything. Thanks!

**Comment by soma-p** [comment]
Added validation + onSubmit handler to the Landlord Form. Tested by running the front-end. Pending items: code for the building of the initial and final forms. Password code completed, to be added once initial and final forms HTML and styled components are complete.

**Comment by YashRavipati1** [comment]
Not sure if others are facing the same issue or if this is just because the branch has not been merged with main in a while, but I can't seem to properly login to check the actual page. Everytime I try to login it just redirects me to the login page again

**Comment by YashRavipati1** [comment]
> Not sure if others are facing the same issue or if this is just because the branch has not been merged with main in a while, but I can't seem to properly login to check the actual page. Everytime I try to login it just redirects me to the login page again

So it seems to be trying to create a new user on the staging environment everytime I try to login. Could be something to look into?

File: frontend/src/App.tsx [hunk 1]
```diff
@@ -6,24 +6,34 @@ import { ThemeProvider } from "styled-components";
...
+  LandlordListingForm,
+  Login,
+  Profile,
+  UnitDetails,
+} from "@/pages";
 import { GlobalStyle, theme } from "@/theme";
 
 function AppRouter() {
   const { signedIn } = useContext(AuthContext);
+  console.log(signedIn);
```
Comment (YashRavipati1) [inline frontend/src/App.tsx]: Make sure to remove this

File: frontend/src/contexts/AuthContext.tsx [hunk 2]
```diff
@@ -35,10 +35,7 @@ export function AuthProvider({ children }: ProviderProps) {
     return unsubscribe;
   }, []);
 
-  const value = useMemo(
-    () => ({ currentUser, signedIn: currentUser !== undefined, loading }),
-    [currentUser],
-  );
+  const value = useMemo(() => ({ currentUser, signedIn: !currentUser, loading }), [currentUser]);
```
Comment (YashRavipati1) [inline frontend/src/contexts/AuthContext.tsx]: Think the issue with sign in could be here, as I believe you're setting signedIn to true if the currentUser doesn't exist and false if it does.

**Review by YashRavipati1 — COMMENTED** [review]
Quick change to auth context to ensure we can actually see the pages. I can test by just making the change myself as well, but thought you should be aware of the issue.

File: frontend/src/contexts/AuthContext.tsx (hunk 2 above)
Comment (petabite) [reply frontend/src/contexts/AuthContext.tsx]:
```suggestion
  const value = useMemo(() => ({ currentUser, signedIn: !!currentUser, loading }), [currentUser]);
```

Ah good catch. This should fix it

**Review by petabite — APPROVED** [review]
🚀

**Pull request merged**